### ISSUE TITLE
broken tab keyboard focus order when editing data

### ISSUE THREAD

**Issue by qwesda** [jakob/Postico#1]
when editing the data in a given row the `tab` key selects the next column as is expected. `shift + tab` moves the focus to the last column instead of selecting the previous column

**Comment by jakob** [comment]
`Shift-tab` seems to work fine for me. Does it happen in every table, or just in same tables?

What version of OS X are you on?

**Comment by qwesda** [comment]
hmm, it seems to work for some columns. but it definitely fails once the keyboard focus is in the last column. It also tends to work for the first 3 columns but not for later ones.
I'm on 10.10.3

**Comment by qwesda** [comment]
the error doesn't occur in tables with only 3 column

**Comment by jakob** [comment]
I can't find a way to reproduce this problem. Tabbing forward and backward works just fine for every table I try. It's also an odd problem, because the tab commands are handled by the system-provided NSTableView.

Just to make sure, you are talking about moving between cells in the table data view, right?

Could you try creating the following table and see if it also exhibits the problem? Maybe the problem is related to your specific tables (eg. data types).

```sql
create table tab_test (col1 text, col2 text, col3 text, col4 text)
```

**Comment by qwesda** [comment]
it can be reproduced with NULL values in columns. try the following setup:

```sql
CREATE SCHEMA tab;

CREATE TABLE tab.tab_test3 (col1 text, col2 text, col3 text);
CREATE TABLE tab.tab_test4 (col1 text, col2 text, col3 text, col4 text);
CREATE TABLE tab.tab_test5 (col1 text, col2 text, col3 text, col4 text, col5 text);

INSERT INTO tab.tab_test3 (col1, col2, col3) VALUES ('col1', 'col2', 'col3');
INSERT INTO tab.tab_test3 (col1, col2, col3) VALUES ('col1', NULL,   'col3');
INSERT INTO tab.tab_test3 (col1, col2, col3) VALUES (NULL,   'col2', 'col3');
INSERT INTO tab.tab_test3 (col1, col2, col3) VALUES (NULL,   NULL,   'col3');
INSERT INTO tab.tab_test3 (col1, col2, col3) VALUES ('col1', 'col2', NULL);
INSERT INTO tab.tab_test3 (col1, col2, col3) VALUES ('col1', NULL,   NULL);
INSERT INTO tab.tab_test3 (col1, col2, col3) VALUES (NULL,   'col2', NULL);
INSERT INTO tab.tab_test3 (col1, col2, col3) VALUES (NULL,   NULL, NULL);

INSERT INTO tab.tab_test4 (col1, col2, col3, col4) VALUES ('col1', 'col2', 'col3', 'col4');
INSERT INTO tab.tab_test4 (col1, col2, col3, col4) VALUES ('col1', NULL,   'col3', 'col4');
INSERT INTO tab.tab_test4 (col1, col2, col3, col4) VALUES (NULL,   'col2', 'col3', 'col4');
INSERT INTO tab.tab_test4 (col1, col2, col3, col4) VALUES (NULL,   NULL,   'col3', 'col4');
INSERT INTO tab.tab_test4 (col1, col2, col3, col4) VALUES ('col1', 'col2', NULL,   'col4');
INSERT INTO tab.tab_test4 (col1, col2, col3, col4) VALUES ('col1', NULL,   NULL,   'col4');
INSERT INTO tab.tab_test4 (col1, col2, col3, col4) VALUES (NULL,   'col2', NULL,   'col4');
INSERT INTO tab.tab_test4 (col1, col2, col3, col4) VALUES (NULL,   NULL,   NULL,   'col4');
INSERT INTO tab.tab_test4 (col1, col2, col3, col4) VALUES ('col1', 'col2', 'col3', NULL);
INSERT INTO tab.tab_test4 (col1, col2, col3, col4) VALUES ('col1', NULL,   'col3', NULL);
INSERT INTO tab.tab_test4 (col1, col2, col3, col4) VALUES (NULL,   'col2', 'col3', NULL);
INSERT INTO tab.tab_test4 (col1, col2, col3, col4) VALUES (NULL,   NULL,   'col3', NULL);
INSERT INTO tab.tab_test4 (col1, col2, col3, col4) VALUES ('col1', 'col2', NULL,   NULL);
INSERT INTO tab.tab_test4 (col1, col2, col3, col4) VALUES ('col1', NULL,   NULL,   NULL);
INSERT INTO tab.tab_test4 (col1, col2, col3, col4) VALUES (NULL,   'col2', NULL,   NULL);
INSERT INTO tab.tab_test4 (col1, col2, col3, col4) VALUES (NULL,   NULL,   NULL,   NULL);

INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES ('col1', 'col2', 'col3', 'col4', 'col5');
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES ('col1', NULL,   'col3', 'col4', 'col5');
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES (NULL,   'col2', 'col3', 'col4', 'col5');
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES (NULL,   NULL,   'col3', 'col4', 'col5');
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES ('col1', 'col2', NULL,   'col4', 'col5');
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES ('col1', NULL,   NULL,   'col4', 'col5');
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES (NULL,   'col2', NULL,   'col4', 'col5');
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES (NULL,   NULL,   NULL,   'col4', 'col5');
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES ('col1', 'col2', 'col3', NULL,   'col5');
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES ('col1', NULL,   'col3', NULL,   'col5');
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES (NULL,   'col2', 'col3', NULL,   'col5');
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES (NULL,   NULL,   'col3', NULL,   'col5');
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES ('col1', 'col2', NULL,   NULL,   'col5');
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES ('col1', NULL,   NULL,   NULL,   'col5');
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES (NULL,   'col2', NULL,   NULL,   'col5');
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES (NULL,   NULL,   NULL,   NULL,   'col5');
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES ('col1', 'col2', 'col3', 'col4', NULL);
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES ('col1', NULL,   'col3', 'col4', NULL);
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES (NULL,   'col2', 'col3', 'col4', NULL);
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES (NULL,   NULL,   'col3', 'col4', NULL);
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES ('col1', 'col2', NULL,   'col4', NULL);
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES ('col1', NULL,   NULL,   'col4', NULL);
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES (NULL,   'col2', NULL,   'col4', NULL);
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES (NULL,   NULL,   NULL,   'col4', NULL);
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES ('col1', 'col2', 'col3', NULL,   NULL);
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES ('col1', NULL,   'col3', NULL,   NULL);
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES (NULL,   'col2', 'col3', NULL,   NULL);
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES (NULL,   NULL,   'col3', NULL,   NULL);
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES ('col1', 'col2', NULL,   NULL,   NULL);
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES ('col1', NULL,   NULL,   NULL,   NULL);
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES (NULL,   'col2', NULL,   NULL,   NULL);
INSERT INTO tab.tab_test5 (col1, col2, col3, col4, col5) VALUES (NULL,   NULL,   NULL,   NULL,   NULL);
```

**Comment by jakob** [comment]
I tried this, but it still seems to work perfectly for me.

Could you provide exact step-by-step instructions or a screen video recording of the bug? (You can record the screen with quicktime player.)

**Comment by qwesda** [comment]
I sent you the screen recording per mail. Github doesn't seem like arbitrary file attachments on issues

**Comment by jakob** [comment]
Thanks for the screen recording. It looks like Shift-Tab doesn't work when you are in an empty text field. It works perfectly on my Mac (OS X 10.10.3).

Do you have any 3rd party tools installed that could interfere with tab? Any custom keyboard shortcuts or hot keys including tab?

**Comment by qwesda** [comment]
ok ... I created a new user on my system, set up a fresh postgres install, etc. ... and the bug did not occur. After a while of trial and error I found the setting that introduces the bug: `preferences.app > keyboard > full keyboard access: "all controls"`.

Try turning it on on your system.

**Comment by jakob** [comment]
Thanks, I can reproduce it now. I have to look into this.

**Comment by jakob** [comment]
I managed to track down the issue and fix the underlying problem. Thanks again for your detailed error reports and patience in tracking this issue down!

The problem will be fixed in the next version.
